### DISCUSSION TITLE
How to know the meaning of dimensions of MPAS-O output

### DISCUSSION THREAD

**Discussion by xiangli1025** [E3SM-Project/E3SM#6619] [E3SM Data help] [answered]
Hello,

Here is a list of dimensions of a MPAS-O output file:

![image](https://github.com/user-attachments/assets/1a9cbdb0-4f58-431a-a37a-f8eed49af7bf)

However, there are no variables such as nRegions, nVertLevels, or nMocStreamfunctionBinsP1.

For example, I can only see nRegions=5, but how can I know the meaning of each nRegions?

Thanks,
Xiang

**Comment by milenaveneziani** [comment, marked as answer]
@xiangli1025: apologies, that is really something missing from our output files. 
The `nRegions` refers to the regions where the MOC streamfunction is calculated, and they are:
```
regionNames = 
    "Atlantic_MOC", 
    "AtlanticMed_MOC", 
    "IndoPacific_MOC", 
    "Pacific_MOC", 
    "Indian_MOC"
```
This can be found in the `$meshname_mocBasinsAndTransects` file, in the relevant inputdata directory. So, for example, the Atlantic_MOC region extends from the Arctic to 34S and does *not* include the Mediterranean. 

The `nOceanRegions` refer to something different, as they represent specific regions where certain quantities are calculated online as the model is running. They include the El Nino regions, for example. I will have to go into the code to tell you exactly what these are, but if you are interested I can do that.

**Reply by xiangli1025** [reply]
Hi Milena,

Thanks for your reply!

May I know the values of nVertLevels and nMocStreamfunctionBinsP1 as well? I'm going to plot the AMOC streamfunction, so these should be necessary. Did you mean I can find the information in the inputdata directory? But I'm not sure where it is.. Could you give me an example of how to find that?

Thanks,
Xiang

**Comment by milenaveneziani** [comment]
`nRegions`, `nVertLevels`, and `nMocStreamfunctionBinsP1` are just dimensions. The first refers to the regions mentioned above, the second is the vertical level dimension, and the third one is the number of latitudinal bins for which the MOC streamfunction is calculated. Since it's 181, it means that the streamfunction is calculated for 1degree latitude bins.
If you want to plot the MOC streamfunction, this is the variable to consider:
```
float timeMonthly_avg_mocStreamvalLatAndDepthRegion(Time, nRegions, nVertLevels, nMocStreamfunctionBinsP1) ;
		timeMonthly_avg_mocStreamvalLatAndDepthRegion:units = "sverdrups" ;
		timeMonthly_avg_mocStreamvalLatAndDepthRegion:long_name = "The value of the MOC streamfunction for each latitude-bin (first dimension) and depth (second dimension)"
```
in the `timeSeriesStatsMonthly` output files. If you want to plot a MOC climatology for the region 1 (the Atlantic without the Mediterranean), then you calculate a time average of `timeMonthly_avg_mocStreamvalLatAndDepthRegion` for nRegion=1, and you will get a function of depth and latBin. The latBin variable is called `binBoundaryMocStreamfunction`, and it's in radians, so you need to multiply it by 180/pi in order to get degrees.

Hope this helps.

**Reply by xiangli1025** [reply]
Hi Milena,

Thanks for your helpful explanation! Now I can plot the AMOC streamfunction averaged over 1850-2014 using one historical member. Please see the figure below. I regarded binBoundaryMocStreamfunction as a coordinate variable for dimension nMocStreamfunctionBinsP1. I wonder if there is a coordinate variable for dimension nVertLevels? Is the unit of the vertical axis a hundred meters?

![image](https://github.com/user-attachments/assets/e4206691-456a-46d1-84fa-97532f281252)


Thanks,
Xiang

**Reply by milenaveneziani** [reply]
That looks great, Xiang!
You can use `refBottomDepth`, which is the mean vertical layer depth. Since this is a constant field, it is not part of the monthly output, but can be found in the mesh file (we also have `layerThickness`, which is the evolving mean thickness of each vertical layer, but you don't need to recalculate the layer depths at every month for just plotting the MOC climatology). 
Do you know how to find MPAS meshfiles? I assume you are looking at v2.1 or v2.0 low-res simulations? In that case, the mesh file to use is one of the `mpaso.EC30to60E2r2*.nc` or `ocean.EC30to60E2r2*nc` files in here: https://web.lcrc.anl.gov/public/e3sm/inputdata/ocn/mpas-o/EC30to60E2r2/
BTW, that folder also contains the `EC30to60E2r2_mocBasinsAndTransects20210623.nc` file that I was telling you about in the first comment. 
Make sure that that is the right mesh name (EC30to60E2r2), though, otherwise things won't work properly.

**Reply by xiangli1025** [reply]
Hi Milena,

Yes, that really helps! I added the refBottomDepth as a coordinate variable, and the streamfunction now looks pretty good.

![image](https://github.com/user-attachments/assets/ad740669-3673-4e0b-92ef-0b102d22b4d3)

I also checked the .nc file which contains the description of ocean regions.

Thanks for providing such helpful information, and your timely response!

Best,
Xiang

**Comment by milenaveneziani** [comment]
> Is the unit of the vertical axis a hundred meters?

No, `nVertLevels` is just a dimension (index of the vertical levels). On the other hand, `refBottomDepth` is a function of `nVertLevels`, but it's mean depth in meters.
